src/osd/modules/render/drawbgfx.cpp

```diff
@@ -513,9 +540,22 @@ static void *sdlNativeWindowHandle(SDL_Window *window)
 	case SDL_SYSWM_COCOA:
 		return wmi.info.cocoa.window;
 #endif
-#if defined(SDL_VIDEO_DRIVER_WAYLAND) && SDL_VERSION_ATLEAST(2, 0, 16)
+#if defined(SDL_VIDEO_DRIVER_WAYLAND) && SDL_VERSION_ATLEAST(2, 0, 16) && defined(SDLMAME_USE_WAYLAND)
 	case SDL_SYSWM_WAYLAND:
-		return wmi.info.wl.egl_window;
+		{
+			wl_egl_window *win_impl = (wl_egl_window*)SDL_GetWindowData(window, "wl_egl_window");
+			if(!win_impl)
+			{
+				int width, height;
+				SDL_GetWindowSize(window, &width, &height);
+				struct wl_surface* surface = wmi.info.wl.surface;
+				if(!surface)
+					return nullptr;
+				win_impl = wl_egl_window_create(surface, width, height);
+				SDL_SetWindowData(window, "wl_egl_window", win_impl);
+			}
+			return (void*)win_impl;
```

Comment:
This is non-trivial copy/pasta – shouldn’t it be factored out?

Reply:
Done.